### DISCUSSION TITLE
Main Discussions

### DISCUSSION THREAD

**Discussion by AliensWithdrawPeople** [caenHV/caen_tools#14] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by AliensWithdrawPeople** [comment]
@nikita-p Маякни, как поправишь (мои) косяки.
Мне на своей машине затруднительно тестить, поэтому много багов проскакивает.
Если ты сегодня в 6:10 pm будешь ещё в институте и будет силы/желание, то можно обсудить сделанные изменения и планы по тестам на железе и временные планы выпуска EOL для v2.

**Reply by nikita-p** [reply]
да, ок, в 6:10 буду в инст-е, подходи

**Reply by nikita-p** [reply]
ок, на первый взгляд сейчас в dev ветке рабочая версия, можно от неё дальше двигаться, 
наверное удобнее будет, если ты синхронизуешь свою ветку с dev и продолжишь развивать свою ветку, а я потом эти наработки смёрджу с dev (или ты сам, если появится возможность протестить перед этим)

**Reply by AliensWithdrawPeople** [reply]
Ок.

**Comment by AliensWithdrawPeople** [comment]
@nikita-p Привет! Я накатил на dq11 для тестов версию из dev (см. [tag v2.3.8-dev](https://github.com/caenHV/caen_tools/releases/tag/v2.3.8-dev)).
Поведение при падении не проверял, но проверил систему логирования и попутно чтение из чтение интерлока с сокета:
* Логи работают, в файлик для ODB пишутся, в нашу базу тоже. На фронтенд выводится без проблем. 
У меня только есть просьба: можешь на страницу с логгами выводить время выдачи, то есть момент загрузки лога. Они же там сами не обновляются. Из-за этого может быть непонятна актуальность загруженной страницы. Это не к спеху. Я в репос с фронтом добавил [issue](https://github.com/caenHV/frontend_webpage/issues/6).
* Интерлок с сокета читается достаточно быстро. Обычно весь скрипт с интерлоком занимает 0.1-0.3 секунды, но иногда время выполнения вырастает до 2 с чем-то секунд. В принципе это ок, так как достаточно редко происходит.

Завтра я поставлю полное напряжение и буду проверять полноценную работу, в том числе и поведение при экстренном падении. По итогам дня я напишу инструкции для дежурных.

Как я понял ты сейчас занялся рефакторингом. То есть в принципе версия v2.3.8-dev является нашим полностью рабочим (с точностью до багов и т.п.) срезом на данный момент. Согласен?

**Reply by nikita-p** [reply]
Да, хорошо, давай от v2.3.8-dev сейчас стартовать, фиксану веб, а рефактор не буду вмёрдживать в dev в ближайшее время
(там сейчас перенёс лоадер в монитор и хочу чек чуть разгрести, но это уже в более размеренном режиме позднее)

оставлю здесь, чтоб не потерять: оказывается, пока мы всё это пилим, вышел официальный python wrapper от caen [github.com/caenspa/py-caen-libs](https://github.com/caenspa/py-caen-libs)

**Reply by AliensWithdrawPeople** [reply]
> оставлю здесь, чтоб не потерять: оказывается, пока мы всё это пилим, вышел официальный python wrapper от caen [github.com/caenspa/py-caen-libs](https://github.com/caenspa/py-caen-libs)

Неожиданно. Позже попытаюсь переписать Setup через эту либу. Будет хорошим поводом для рефакторинга, а то сейчас там какой-то ужас в коже нагорожен.

Спасибо за наводку. Если не секрет, ты как это нашёл?

**Reply by nikita-p** [reply]
Случайно практически: искал caen_tools в глобальном поиске по коду на GitHub, там в топе оказался официальный репос

**Comment by nikita-p** [comment]
Залил новый релиз v2.3.9-dev с мелкими улучшениями и фиксами фронтенда (по большей части) и добавил немного отсылок статусов от контролей в мон.логгер

https://github.com/caenHV/caen_tools/releases/tag/v2.3.9-dev

**Comment by nikita-p** [comment]
Новый день - новые обновления

Для алармов нужно дописать дополнительных полей в last_measurement.json (сейчас добавил поле AUTOPILOT 0/1 по просьбе АСПопова), можно подумать, что ещё надо

Плюс просят как минимум скопировать код в гитлаб ияфа (тебе должны были уже создать аккаунт), я скопировал актуальные версии репосов, и ещё хотят, чтоб весь процесс релиза был более быстрым

**Reply by AliensWithdrawPeople** [reply]
> Для алармов нужно дописать дополнительных полей в last_measurement.json (сейчас добавил поле AUTOPILOT 0/1 по просьбе АСПопова), можно подумать, что ещё надо

Для статусов уже есть отдельный файлик. Имхо, так удобнее писать и считывать, учитывая то, что статусы и напряжения не вместе пишутся.

> Плюс просят как минимум скопировать код в гитлаб ияфа (тебе должны были уже создать аккаунт), я скопировал актуальные версии репосов, и ещё хотят, чтоб весь процесс релиза был более быстрым

Bruh... Что значит более быстрым? Что не так с нынешним воркфлоу?

**Reply by nikita-p** [reply]
> Bruh... Что значит более быстрым? Что не так с нынешним воркфлоу?

Я принимаю главный аргумент, что не круто пересобирать весь образ с нуля, если мы обновили свой пакет
(Я сегодня добавил тэг образу base (теперь в docker images он не ноунейм образ) и вроде теперь он не пересобирается, а берётся из кэша, может это уже полдела)

Были ещё тезисы, но их лучше как-нибудь потом лично обсудить

**Reply by AliensWithdrawPeople** [reply]
А можешь ссылку дать на гитлаб этот? Мне ни на ияфовскую, ни на мою личную почту ничего не приходило.

**Reply by nikita-p** [reply]
да, вот например https://git.inp.nsk.su/cmd3/caendc/tools

подробности у Попова наверное можно выяснить, т.к. он вчера писал письмо админу, который должен регать; 
но да, в поиске по людям я тебя пока не вижу

**Reply by AliensWithdrawPeople** [reply]
Спасибо! Правда, без аутентификации я ничего не вижу. Мои стандартные пароли от ияфовской почты не проходят. Буду с Поповым разбираться.

**Comment by AliensWithdrawPeople** [comment]
Я тут немного переписал всю часть с проверкой состояний. Раньше была жуть из кучи if-else фигни, в которой проследить логику было сложно. Теперь я всё это разбил на относительно небольшие логические части. 
Из удобного --- я сделал датакласс ChannelStatus, который +- всеобъемлюще описывает состояние канала. Мне показалось это очень удобным. В будущем при имлементации #23 такой формат передачи будет удобнее. 
Там ещё по мелочи изменил ErrorCode. Пока это не существенно, но опять же при разбиении HealthCheck такого рода изменения станут QOL улучшением, imo.

К тому же, теперь трип тайм распространяется и на оверкаррент, согласно запросу Попова. 


Все эти изменения в [9a6f3e9](https://github.com/caenHV/caen_tools/commit/9a6f3e9b4568b8b0b79554c50084894c8a34c587). 
Я эту лабуду ещё не потестил, но сразу в дев залил, не подумав. Сорри.

**Reply by nikita-p** [reply]
хорошо, я на фейке чекну

**Comment by AliensWithdrawPeople** [comment]
Попов требует переработать Reducer:

> Массив напряжений для махания раз в полчаса на 20 секунд. Ключик для включения режима лучше можно передавать через конфиг

Коэффциент снижения считать по этой формуле:
![image](https://github.com/user-attachments/assets/6912eff7-7563-4370-8d6f-f3592b43a39b)

![image](https://github.com/user-attachments/assets/1b4a9fac-3add-4cfc-8cad-ffbda7105380)

Говорит, что в его программе так и работает.
Я не понимаю, почему он сразу этого не сказал. 

> Ключик для включения режима лучше можно передавать через конфиг

Это и так у нас есть через enable. Позже ему это объясню.

**Comment by nikita-p** [comment]
@AliensWithdrawPeople 

релиз v2.3.10-dev (теперь редьюсер можно убирать, для этого в конфиге autopilot/reduce нужно убрать его упоминание)
https://github.com/caenHV/caen_tools/releases/tag/v2.3.10-dev

**Reply by AliensWithdrawPeople** [reply]
@nikita-p Привет! 
Сейчас на dq11 стоят вот такие параметры в конфиге. Несмотря на это в 15:39 запустился reducer. Я что-то не так сделал?
```
[check.interlock]
;Interlock settings (reads interlock to have actual values for MChS)
enable = true
; Period of the interlock checks, s
repeat_every = 1

[check.autopilot]
;Autopilot settings
run = relax
; If interlock is up, voltage_multiplier = target_voltage * voltage_modifier
voltage_modifier = 0.4
; As a multiplier of a base voltage (check out map_config)
target_voltage = 1

[check.autopilot.relax]
;Relax settings
repeat_every = 15
voltage_modifier = ${check.autopilot:voltage_modifier}
target_voltage = ${check.autopilot:target_voltage}

[check.autopilot.reducer]
;Scheduled reducer settings
; Every repeat_every (s) voltage is temporarily reduced for reducing_period (s)
repeat_every = 1800
reducing_period = 60
voltage_modifier = ${check.autopilot:voltage_modifier}
target_voltage = ${check.autopilot:target_voltage}
```

**Reply by nikita-p** [reply]
сейчас посмотрю

**Reply by nikita-p** [reply]
нашёл проблему 

HealthControl сейчас поднимает все dependent scripts, если после пробоя всё норм
https://github.com/caenHV/caen_tools/blob/eeeac62f4b97489849496439c822bc33148a9a59/caen_tools/SystemCheck/scripts/health.py#L281

подтверждается косвенно логами, за полчаса до редьюсера было это 

> [2024-10-30 15:09:02] WARNING in health.py:141:: Channel 1 is not in a ramp down but is in either over/under voltage or over current. It is on its last breath trip time.
> [2024-10-30 15:09:02] WARNING in health.py:164:: Something went wrong while the trip time criteria had been checked: not enough values to unpack (expected 2, got 1)
> [2024-10-30 15:09:02] INFO in health.py:187:: Ramping status detected
> [2024-10-30 15:09:02] WARNING in health.py:198:: Device is not ok: error = ErrorCode.CANT_READ, description = Something went wrong while the trip time criteria had been checked.
> [2024-10-30 15:09:02] ERROR in health.py:214:: Bad deivce parameters. Emergency DownVoltage!
> [2024-10-30 15:09:02] WARNING in metascript.py:46:: Stop the script
> [2024-10-30 15:09:02] INFO in mchswork.py:73:: Send NACK to MChS
> [2024-10-30 15:09:02] INFO in metascript.py:83:: Task was cancelled
> [2024-10-30 15:09:03] INFO in health.py:224:: Send emergency status of the device: is_ok = False, description = Something went wrong while the trip time criteria had been checked.
> [2024-10-30 15:09:03] INFO in health.py:345:: HealthControl was done in 1.025 s
 

а запуск syscheck последний перед редьюсером был в 14:22 (т.е. больше получаса перед срабатыванием)
> [2024-10-30 14:22:55] INFO in check.py:39:: Start SysCheck with arguments

**Reply by nikita-p** [reply]
@AliensWithdrawPeople я точно не знаю, как тут лучше переработать код, 
мб просто пусть healthcontrol шлёт на api ресипты (выключи автопилот, включи автопилот), и тогда не пытается самостоятельно кого-то запустить

**Reply by AliensWithdrawPeople** [reply]
Согласен.

**Reply by nikita-p** [reply]
ты фиксанёшь или я?

**Reply by AliensWithdrawPeople** [reply]
Фиксану.

**Reply by nikita-p** [reply]
> (выключи автопилот, включи автопилот)

сейчас осознал, запишу здесь, чтоб не забыть; завершать всё-таки нужно напрямую (поскольку важна последовательность: сначала завершить зависимости, потом снижать; а отправка на api установит только нужные shared_parameters, которые в течение секунды подхватит manager)
а с включением непонятно

**Reply by nikita-p** [reply]
ещё пофиксь плиз эту строку
https://github.com/caenHV/caen_tools/blob/eeeac62f4b97489849496439c822bc33148a9a59/caen_tools/SystemCheck/scripts/health.py#L241

ты норм описания сейчас пишешь в `status.failure[1]`, давай их отправлять, вместо того, что сейчас

**Reply by AliensWithdrawPeople** [reply]
Ок.

**Reply by AliensWithdrawPeople** [reply]
Фиксанул, но как будет работать время покажет.

Ещё я решил вопрос с наками, с которым тебя донимал Попов. Там при выставлении timeout 2500 у MChS  контроллера происходи какй-то косяк. Мы откатили на 2000, но период HealthCheck я уменьшил до 0.5 секунд. 
Там ещё пара косяков была с моего рефакторинга perform_cheks. Я их исправил и залил версию v2.3.10-dev3. На dq11 сейчас она стоит.

**Comment by AliensWithdrawPeople** [comment]
В [v2.3.11-alpha1](https://github.com/caenHV/caen_tools/releases/tag/v2.3.11-alpha1) фиксанул косяки в логиеке перезапуска. 
На dq11 поставил refresh_time = 0, потому что чтение кешированных плохих значений приводит к проблемам с перезапуском, вроде. Хотя надо тестить после фикса логики перезапуска.

**Comment by AliensWithdrawPeople** [comment]
1) @nikita-p Фиксани проблему с повторяющимися логгами, пожалуйста. На фронтенд передавать 20МБ с учётом сети очень плохо. Страничка очень долго загружается. Это нужно прям asap.

2) Что-то вообще срань с залипающими статусами. Я добавил отправку reset тикета в failure_action после down. Но мне кажется это не поможет. Есть ощущение, что какое-то время некоторые статусы не снимаются. То есть всё может быть так, как ты говорил. Если это так,  то я хз, что делать.

**Reply by nikita-p** [reply]
ок, первое сейчас фиксану, про второе подумаю

**Reply by nikita-p** [reply]
залил обновление монитора на dq11, новые статусы должны склеиваться

**Reply by AliensWithdrawPeople** [reply]
Чекнул изменения.  Штука с прагмами реально помогает? Если да, то вообще огонь!
Пока такой вариант ок, но как будет время внесу очередные изменния в логику перезапуска. Сейчас объясню зачем в булетпойнтах: 
* При автоперезапуске чекается количество экстренных падений за последние n секунд по монитору
* При склеивании без пометок необходимая инфа теряется
* Сейчас это должно работать из-за некоторого файнтюна периодов склеивания и автоперезапуска, но это не хорошо
Что я предлагаю сдделать (и сам сделаю, скорее всего): Можно добавить в таблицу status поля, которое будет отображать размер стака. И соответсвенно считать количество экстренных даунов с учётом склеивания.
Что думаешь?

**Reply by nikita-p** [reply]
я бы наверное унёс логику в syscheck
(либо в API, либо даже просто в Health где-то считал сколько падений, мне кажется это проще будет, чем работать с базой)

но в целом я не против того, что ты предлагаешь, ещё погуглю, как это можно реализовать, напишу сюда

**Reply by nikita-p** [reply]
> Я добавил отправку reset тикета в failure_action после down. Но мне кажется это не поможет. 

а если добавить в конце failure_actions периодически повторяющиеся попытки reset до тех пор пока не отлипнет?

> я бы наверное унёс логику в syscheck

предложил вариант в [пулл реквесте](https://github.com/caenHV/caen_tools/pull/25), посмотри плиз, я проверил сам класс, вроде ок, но уже не особо соображаю

**Reply by AliensWithdrawPeople** [reply]
Я замерджил.

**Comment by AliensWithdrawPeople** [comment]
Тут Попов новую пачку требований выкатил:

> Для Даниила Иванова - в HVmonitor требуется:
> 1) ввести допуск на время превышения софтового предела по току (дать возможность
> самому затухнуть)
> 2) Если диагностирован пробой и принято решение снизить напряжение, cнижать
> с рабочего до множителя 0.8 (можно побыстрее). Проверить, есть ли аварийные токи;
> если да, то снижать до нуля, если аварийных токов нет - ждем 5 секунд и ворачиваем
> напряжение до штатного (тоже быстро), разблокируем МЧС и начинаем писать.
> 3) если в течении 1 минуты таких пробоев произошло 5 штук, то снижаем напряжение
> до 0.4, ждем 2-3 минуты, и пытаемся поднять напряжение. /aspopov1/

Вроде, все инструменты для реализации уже есть. Попробую это сделать на выходных.

**Reply by nikita-p** [reply]
хочется надеяться, что это уже завершающий набор требований, если что, пиши, 
я пока в ветке refactor продолжу разбираться

**Reply by AliensWithdrawPeople** [reply]
Я в ветке dev_issue26 начал делать. Если, что я там в issue #26  пару слов об этом всём написал. Чекни, как сможешь, пожалуйста.

**Reply by nikita-p** [reply]
@AliensWithdrawPeople 
привет, у нас всё выключилось в 12:21 и спит, это по плану?

> [2024-11-05 12:21:30] INFO in ws.py:123:: Start server shutdown actions

если нет, то попробую перезапустить

**Reply by AliensWithdrawPeople** [reply]
Не знаю. Я ничего не делал. Попов ничего не сообщал.

**Reply by AliensWithdrawPeople** [reply]
Там device_bacjend отвалился в 00:25:46 (смотри в лог за вчерашний день). После того, как юзер поставил напряжение 0.9.

**Reply by AliensWithdrawPeople** [reply]
Точнее ws не мог подцепиться к devback.

**Reply by nikita-p** [reply]
сегодня утром я заходил на сайт, там кто-то игрался с напряжением, вроде всё ок было; 
пойду Попова поищу спрошу

**Reply by nikita-p** [reply]
1) честно попытался найти Попова в кабинете и на пультовой, но не нашёл 

читаю в еLog:
>  10:40 Снято напряжение с mu системы. Сняли с детектора мюонные счётчики 21-24 для доступа к ДК и ксенону - будут производиться работы.

наверное Попов всё отключил, потому что сам докер так бы не смог, а больше некому

----------------
2) 

> Точнее ws не мог подцепиться к devback.

после отключения кэша и опроса в 0.5с я стал замечать моргающие Device: Fail статусы на фронтнеде, он видимо на пределе сейчас

**Reply by AliensWithdrawPeople** [reply]
> после отключения кэша и опроса в 0.5с я стал замечать моргающие Device: Fail статусы на фронтнеде, он видимо на пределе сейчас

Да, я такое уже пару недель вижу на железе при тестах, но вроде не мешает.

> наверное Попов всё отключил, потому что сам докер так бы не смог, а больше некому

Достаточно странно, но ладно. Кстати, елог так и не отправили на почту. Попов сейчас в ИЯФе?

**Reply by nikita-p** [reply]
> Попов сейчас в ИЯФе?

да, по базе не выходил со вчерашнего вечера

> вижу на железе при тестах

сейчас измерил время исполнения тикета GetParams в Setup, 
есть ощущение, что можно ускорить https://github.com/caenHV/Setup/issues/25

**Comment by nikita-p** [comment]
@AliensWithdrawPeople 
привет, я сейчас перевожу проекты и пайплайны сборки в институтский гитлаб;
хотел тебя добавить, но не вижу профиля, тебя всё ещё там не зарегистрировали?

**Reply by AliensWithdrawPeople** [reply]
Да, меня там нет. Добавлять не надо, поскольку я больше не работаю над проектом.
Если нужно будет что-то объяснить по коду, который я написал, обращайся.

**Reply by nikita-p** [reply]
:cry: 
без тебя будет сложно :(
как бы то ни было, спасибо, было приятно с тобой работать

(насчёт проги, видимо всё как обычно медленно двигается, но двигается:
не знаю, приходило ли тебе, мне тётя из патентного отдела писала пару недель назад, что доки отправили в москву,
так что поди к новому году с этим закончат)

**Reply by AliensWithdrawPeople** [reply]
Мне с тобой тоже было приятно работать. Многому у тебя научился.
Прости, что так оставляю работу.
